compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,272 @@
+                                    ValueRange values) {
+  // Replace all OpResults with the given values.
+  SmallVector<Value> replacements;
+  for (OpResult opResult : op->getOpResults()) {
+    Value replacement = values[opResult.getResultNumber()];
+    // If this value is a vector type, it must be converted back to simd.
+    if (isa<VectorType>(replacement.getType())) {
+      auto oldResult = cast<TypedValue<VectorType>>(opResult);
+      // Create a toSIMD op to convert the value back to the simd.
+      rewriter.setInsertionPointAfterValue(oldResult);
+      auto toSIMD = rewriter.create<IREE::VectorExt::ToSIMDOp>(
+          oldResult.getLoc(), oldResult.getType(), replacement);
+      // Clone the layout to the new value.
+      provider->getAnalysis().cloneLayoutInformationToNewValue(
+          oldResult, toSIMD.getResult());
+      // Add to replacements.
+      replacement = toSIMD.getResult();
+    }
+    replacements.push_back(replacement);
+  }
+
+  rewriter.replaceOp(op, replacements);
+}
+
+class DistributionRewriter : public IRRewriter, public RewriterBase::Listener {
+public:
+  DistributionRewriter(MLIRContext *ctx, DenseSet<Operation *> &erasedOps,
+                       SmallVector<Operation *> &worklist,
+                       VectorLayoutAnalysis &analysis)
+      : IRRewriter(ctx), erasedOps(erasedOps), worklist(worklist),
+        analysis(analysis) {
+    setListener(this);
+  }
+
+protected:
+  void notifyOperationRemoved(Operation *op) override { erasedOps.insert(op); }
+
+  void notifyOperationInserted(Operation *op) override {
+    // For now, do nothing.
+    // TODO: Check if the operation can still be distributed and try to
+    // distribute it. Not sure how to check this? Maybe we can check if the
+    // analysis knows about this value and assume that this value needs to be
+    // distributed. This is needed for SIMD -> SIMD rewrites.
+  }
+
+private:
+  // A reference to the set of operations that have been erased.
+  DenseSet<Operation *> &erasedOps;
+  // A reference to the worklist of operations that need to be distributed.
+  SmallVector<Operation *> &worklist;
+  // A reference to the analysis that provides the layout information.
+  VectorLayoutAnalysis &analysis;
+};
+
+VectorDistribution::VectorDistribution(func::FuncOp root,
+                                       VectorLayoutAnalysis &analysis,
+                                       LayoutProvider *provider)
+    : root(root), analysis(analysis), provider(provider) {
+  provider->setAnchorOps();
+  if (failed(analysis.run()))
+    return;
+  LLVM_DEBUG(llvm::dbgs() << "Layout Analysis Completed Successfully :\n");
+  LLVM_DEBUG(analysis.print(llvm::dbgs()));
+}
+
+static bool canDistribute(Operation *op, VectorLayoutAnalysis &analysis) {
+  bool needsDistribution = false;
+  // Check if this operation has any operands with a vector type. If so,
+  // then they need to have a layout.
+  for (Value operand : op->getOperands()) {
```

Comment:
Since the logic is the same can combine these two loops into one:
```suggestion
  auto values = llvm::to_vector_of<Value>(op->getOperands());
  llvm::append_range(values, op->getResults());
  for (Value value : values) {
```

(Note that we can't just call `llvm::concat` and need to materialize operands/results to some non-temporary values to deal with lifetime extension bugs in C++17.)